runtime/jvmti/suspendhelper.cpp

```diff
@@ -35,33 +35,49 @@ suspendThread(J9VMThread *currentThread, jthread thread, UDATA allowNull, UDATA
 	*currentThreadSuspended = FALSE;
 	rc = getVMThread(currentThread, thread, &targetThread, allowNull, TRUE);
 	if (rc == JVMTI_ERROR_NONE) {
-		if (targetThread->publicFlags & J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND) {
-			rc = JVMTI_ERROR_THREAD_SUSPENDED;
-		} else {
-			if (targetThread->publicFlags & J9_PUBLIC_FLAGS_STOPPED) {
-				/* Do not suspend dead threads. Mirrors SUN behaviour. */
-				rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
+#if JAVA_SPEC_VERSION >= 19
+		j9object_t threadObject = (NULL == thread) ? currentThread->threadObject : J9_JNI_UNWRAP_REFERENCE(thread);
+		BOOLEAN isVirtual = IS_VIRTUAL_THREAD(currentThread, threadObject);
+		if (NULL != targetThread)
+#endif /* JAVA_SPEC_VERSION >= 19 */
+		{
+			if (targetThread->publicFlags & J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND) {
+				rc = JVMTI_ERROR_THREAD_SUSPENDED;
 			} else {
-				if (currentThread == targetThread) {
-					*currentThreadSuspended = TRUE;
+				if (targetThread->publicFlags & J9_PUBLIC_FLAGS_STOPPED) {
+					/* Do not suspend dead threads. Mirrors SUN behaviour. */
+					rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
 				} else {
-					currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
-					omrthread_monitor_enter(targetThread->publicFlagsMutex);
-					VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
-					if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
-						while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
-							omrthread_monitor_wait(targetThread->publicFlagsMutex);
+					if (currentThread == targetThread) {
+						*currentThreadSuspended = TRUE;
+					} else {
+						currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
+						omrthread_monitor_enter(targetThread->publicFlagsMutex);
+						VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
+						if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
+							while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
+								omrthread_monitor_wait(targetThread->publicFlagsMutex);
+							}
 						}
+						omrthread_monitor_exit(targetThread->publicFlagsMutex);
+						currentThread->javaVM->internalVMFunctions->internalEnterVMFromJNI(currentThread);
 					}
-					omrthread_monitor_exit(targetThread->publicFlagsMutex);
-					currentThread->javaVM->internalVMFunctions->internalEnterVMFromJNI(currentThread);
+					Trc_JVMTI_threadSuspended(targetThread);
 				}
-				Trc_JVMTI_threadSuspended(targetThread);
 			}
 		}
+#if JAVA_SPEC_VERSION >= 19
+		if (isVirtual) {
```

Comment:
this should be an `else if` since mounted virtual thread will be suspended by the public flag, adding an extra suspend to the vthread state will cause logic issue?

Reply:
also, virtual thread should also check if the thread is alive or not and return `rc = JVMTI_ERROR_THREAD_NOT_ALIVE;` in those cases

Reply:
the for mounted continuation, `getVirtualThreadState` should return the vmthread state which should contain `JVMTI_THREAD_STATE_SUSPENDED` https://github.com/eclipse-openj9/openj9/blob/master/runtime/util/thrinfo.c#L409-L415